File: generator/src/payments.rs
use crate::utils::generate_uuid;
use chrono::NaiveDate;
use rand::Rng;
use serde::Serialize;

#[derive(Debug, Serialize, Clone)]
pub struct CardDetails {
    pub number: String,
    pub expiration: NaiveDate,
}

#[derive(Debug, Serialize, Clone)]
pub struct Wallet {
    pub expiration: NaiveDate,
}

#[derive(Debug, Serialize, Clone)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum PaymentDetails {
    Card(CardDetails),
    Wallet(Wallet),
}

impl PaymentDetails {
    pub fn expiration_date(&self) -> NaiveDate {
        return match self {
            PaymentDetails::Card(card) => card.expiration,
            PaymentDetails::Wallet(wallet) => wallet.expiration,
        };
    }
}

#[derive(Debug, Serialize, Clone)]
pub struct Payment {
    pub details: PaymentDetails,
    pub transaction_id: String,
}

fn parse_exp_date(exp_str: &str) -> NaiveDate {
    let parts: Vec<u32> = {
        exp_str
            .split('/')
            .map(|s| s.parse().unwrap_or(1))
            .collect()
    };
    let (month, year_suffix) = (parts[0], parts[1]);
    let year = 2000 + year_suffix as i32;

    let first_of_next_month = if month == 12 {
        NaiveDate::from_ymd_opt(year + 1, 1, 1).unwrap()
    } 
    else {
        NaiveDate::from_ymd_opt(year, month + 1, 1).unwrap()
    };
    return first_of_next_month.pred_opt().unwrap();
}

fn get_card_list() -> Vec<CardDetails> {
    return vec![
        ("************5171", "05/25"), ("************7690", "05/27"),
        ("************2763", "06/24"), ("************3315", "04/26"),
        ("************4072", "01/28"), ("************6869", "11/25"),
        ("************2781", "08/26"), ("************0904", "09/25"),
        ("************4469", "09/27"), ("************5164", "08/26"),
        ("************3341", "04/25"), ("************1349", "07/26"),
        ("************4167", "03/25"), ("************7168", "09/25"),
        ("************8239", "10/26"), ("************0932", "11/26"),
        ("************8186", "08/27"), ("************7668", "11/25"),
        ("************9486", "09/27"), ("************1093", "02/27"),
        ("************0238", "01/26"), ("************9611", "02/28"),
        ("************7295", "09/26"), ("************8346", "08/27"),
        ("************2464", "01/26"),
    ].into_iter().map(|(n, d)| CardDetails {
        number: n.to_string(),
        expiration: parse_exp_date(d),
    })
    .collect();
}

fn get_wallet_list() -> Vec<Wallet> {
    return {
        get_card_list()
            .iter()
            .map(|card| Wallet { expiration: card.expiration })
            .collect()
    };
}

pub fn choose_payment_method(name: &str) -> PaymentDetails {
    let cards = get_card_list();
    let wallets = get_wallet_list();
    let name_to_index = [
        "Ava Whitaker", "Liam Caldwell", "Isabella Greene", "Ethan Morrell",
        "Maya Ellison", "Noah Blackwood", "Chloe Hartman", "Lucas Pennington",
        "Sofia Langford", "Oliver Drayton", "Harper Linwood", "Sebastian Knox",
        "Amelia Fairbanks", "Julian Royce", "Nora Halston", "Elijah Trent",
        "Zoe Merrick", "Caleb Winslow", "Lily Hargrove", "Milo Carrington",
        "Aria Templeton", "Declan Shore", "Vivian Leclair", "Grayson Holt",
        "Clara Redmond",
    ];
    if let Some(index) = name_to_index.iter().position(|n| *n == name) {
        let mut rng = rand::rng();
        if rng.random_bool(0.5) { PaymentDetails::Card(cards[index].clone()) } 
        else { PaymentDetails::Wallet(wallets[index].clone()) }
    } 
    else { panic!("Unknown customer name: {name}"); }
}

pub fn new_payment(name: &str) -> Payment {
    return Payment {
        details: choose_payment_method(name),
        transaction_id: generate_uuid(),
    };
}

File: past_approaches/old_generator/payments/payments.go
package payments

import (
	"generator/utils"
	"math/rand"
	"time"
)

type PaymentMethod interface {
	ExpirationDate() time.Time
}

type CardDetails struct {
	Number     string
	Expiration time.Time
}

type Wallet struct {
	Expiration time.Time
}

func (c CardDetails) ExpirationDate() time.Time {
	return c.Expiration
}

func (w Wallet) ExpirationDate() time.Time {
	return w.Expiration
}

type Payment struct {
	Details       PaymentMethod
	TransactionId string
}

func parseExpDate(expStr string) time.Time {
	layout := "01/06"
	t, _ := time.Parse(layout, expStr)
	// Add 1 month, then subtract 1 day to get the last day of the original month
	firstOfNextMonth := t.AddDate(0, 1, 0)
	endOfMonth := firstOfNextMonth.AddDate(0, 0, -1)
	return endOfMonth
}

var CardList = map[int64]CardDetails{
	0:  {"************5171", parseExpDate("05/25")},
	1:  {"************7690", parseExpDate("05/27")},
	2:  {"************2763", parseExpDate("06/24")},
	3:  {"************3315", parseExpDate("04/26")},
	4:  {"************4072", parseExpDate("01/28")},
	5:  {"************6869", parseExpDate("11/25")},
	6:  {"************2781", parseExpDate("08/26")},
	7:  {"************0904", parseExpDate("09/25")},
	8:  {"************4469", parseExpDate("09/27")},
	9:  {"************5164", parseExpDate("08/26")},
	10: {"************3341", parseExpDate("04/25")},
	11: {"************1349", parseExpDate("07/26")},
	12: {"************4167", parseExpDate("03/25")},
	13: {"************7168", parseExpDate("09/25")},
	14: {"************8239", parseExpDate("10/26")},
	15: {"************0932", parseExpDate("11/26")},
	16: {"************8186", parseExpDate("08/27")},
	17: {"************7668", parseExpDate("11/25")},
	18: {"************9486", parseExpDate("09/27")},
	19: {"************1093", parseExpDate("02/27")},
	20: {"************0238", parseExpDate("01/26")},
	21: {"************9611", parseExpDate("02/28")},
	22: {"************7295", parseExpDate("09/26")},
	23: {"************8346", parseExpDate("08/27")},
	24: {"************2464", parseExpDate("01/26")},
}

var DigitalWalletList = map[int64]Wallet{
	0:  {parseExpDate("05/25")},
	1:  {parseExpDate("05/27")},
	2:  {parseExpDate("06/24")},
	3:  {parseExpDate("04/26")},
	4:  {parseExpDate("01/28")},
	5:  {parseExpDate("11/25")},
	6:  {parseExpDate("08/26")},
	7:  {parseExpDate("09/25")},
	8:  {parseExpDate("09/27")},
	9:  {parseExpDate("08/26")},
	10: {parseExpDate("04/25")},
	11: {parseExpDate("07/26")},
	12: {parseExpDate("03/25")},
	13: {parseExpDate("09/25")},
	14: {parseExpDate("10/26")},
	15: {parseExpDate("11/26")},
	16: {parseExpDate("08/27")},
	17: {parseExpDate("11/25")},
	18: {parseExpDate("09/27")},
	19: {parseExpDate("02/27")},
	20: {parseExpDate("01/26")},
	21: {parseExpDate("02/28")},
	22: {parseExpDate("09/26")},
	23: {parseExpDate("08/27")},
	24: {parseExpDate("01/26")},
}

func ChoosePaymentMethod(name string) PaymentMethod {
	paymentMethods := map[string][]PaymentMethod{
		"Ava Whitaker":     {CardList[0], DigitalWalletList[0]},
		"Liam Caldwell":    {CardList[1], DigitalWalletList[1]},
		"Isabella Greene":  {CardList[2], DigitalWalletList[2]},
		"Ethan Morrell":    {CardList[3], DigitalWalletList[3]},
		"Maya Ellison":     {CardList[4], DigitalWalletList[4]},
		"Noah Blackwood":   {CardList[5], DigitalWalletList[5]},
		"Chloe Hartman":    {CardList[6], DigitalWalletList[6]},
		"Lucas Pennington": {CardList[7], DigitalWalletList[7]},
		"Sofia Langford":   {CardList[8], DigitalWalletList[8]},
		"Oliver Drayton":   {CardList[9], DigitalWalletList[9]},
		"Harper Linwood":   {CardList[10], DigitalWalletList[10]},
		"Sebastian Knox":   {CardList[11], DigitalWalletList[11]},
		"Amelia Fairbanks": {CardList[12], DigitalWalletList[12]},
		"Julian Royce":     {CardList[13], DigitalWalletList[13]},
		"Nora Halston":     {CardList[14], DigitalWalletList[14]},
		"Elijah Trent":     {CardList[15], DigitalWalletList[15]},
		"Zoe Merrick":      {CardList[16], DigitalWalletList[16]},
		"Caleb Winslow":    {CardList[17], DigitalWalletList[17]},
		"Lily Hargrove":    {CardList[18], DigitalWalletList[18]},
		"Milo Carrington":  {CardList[19], DigitalWalletList[19]},
		"Aria Templeton":   {CardList[20], DigitalWalletList[20]},
		"Declan Shore":     {CardList[21], DigitalWalletList[21]},
		"Vivian Leclair":   {CardList[22], DigitalWalletList[22]},
		"Grayson Holt":     {CardList[23], DigitalWalletList[23]},
		"Clara Redmond":    {CardList[24], DigitalWalletList[24]},
	}
	options := paymentMethods[name]
	choice := rand.Intn(2)
	return options[choice]
}

func NewPayment(name string) Payment {
	method := ChoosePaymentMethod(name)
	return Payment{method, utils.GenerateUuid()}
}
